realm/realm-library/src/main/cpp/io_realm_internal_Collection.cpp

```diff
@@ -0,0 +1,299 @@
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto results = reinterpret_cast<Results*>(native_ptr);
+        return static_cast<jlong>(results->size());
+    } CATCH_STD()
+    return 0;
+}
+
+JNIEXPORT jobject JNICALL
+Java_io_realm_internal_Collection_nativeAggregate(JNIEnv *env, jclass, jlong native_ptr, jlong column_index,
+        jbyte agg_func)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto results = reinterpret_cast<Results *>(native_ptr);
+
+        size_t index = S(column_index);
+        Optional<Mixed> value;
+        switch (agg_func) {
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_MINIMUM:
+                value = results->min(index);
+                break;
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_MAXIMUM:
+                value = results->max(index);
+                break;
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_AVERAGE:
+                value = results->average(index);
+                break;
+            case io_realm_internal_Collection_AGGREGATE_FUNCTION_SUM:
+                value = results->sum(index);
+                break;
+        }
+
+        if (!value) {
+            return static_cast<jobject>(nullptr);
+        }
+
+        Mixed m = *value;
+        switch (m.get_type()) {
+            case type_Int:
+                return NewLong(env, m.get_int());
+            case type_Float:
+                return NewFloat(env, m.get_float());
+            case type_Double:
+                return NewDouble(env, m.get_double());
+            case type_Timestamp:
+                return NewDate(env, m.get_timestamp());
+            default:
+                throw std::invalid_argument("Excepted numeric type");
+        }
+    } CATCH_STD()
+    return static_cast<jobject>(nullptr);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_Collection_nativeSort(JNIEnv *env, jclass, jlong native_ptr, jlong sort_desc_native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+    try {
+        auto results = reinterpret_cast<Results*>(native_ptr);
+        auto sort_descriptor = *reinterpret_cast<SortDescriptor*>(sort_desc_native_ptr);
+        auto sorted_result = results->sort(std::move(sort_descriptor));
+        return reinterpret_cast<jlong>(new Results(std::move(sorted_result)));
+    } CATCH_STD()
+    return reinterpret_cast<jlong>(nullptr);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_Collection_nativeAddListener(JNIEnv* env, jobject instance, jlong native_ptr)
+{
+    TR_ENTER_PTR(native_ptr)
+
+    try {
+        auto results = reinterpret_cast<Results*>(native_ptr);
+
+        // FIXME: Those need to be freed for all the corner cases!
+        jobject weak_results = env->NewWeakGlobalRef(instance);
+
+        auto cb = [=](realm::CollectionChangeSet const& changes,
+                                   std::exception_ptr err) {
+            jclass results_class = env->GetObjectClass(weak_results);
```

Comment:
Maybe we can do the look up outside the callback function `cb` as it might slow down the actual call back.